### PR TITLE
UCT/IB: Use separate resource domain for mlx5 trasnports

### LINKED ISSUE #1926: [mtt] Timeout in mpi_test_suite with HW TM
Configuration:
```
MOFED: MLNX_OFED_LINUX-4.1-4.1.1.0
OMPI: 4.0.0a1
Orion x36 (clx-orion-[022,025,028-032,037-038,045-047,049,053,063-064,066-071,081-090,092-095])
```
MTT: http://e2e-gw.mellanox.com:4080//hpc/scrap/users/mtt/scratch/hcol/20171016_215200_22084_17122_clx-orion-071/html/test_stdout_zAiLea.txt

Cmd:
`mpirun -np 1008 --debug-daemons --display-map --bind-to core --map-by node -mca pml ucx -mca btl_openib_warn_default_gid_prefix 0 -mca btl_openib_if_include mlx5_0:1 --timestamp-output -x HCOLL_IB_IF_INCLUDE=mlx5_0:1 -x MXM_RDMA_PORTS=mlx5_0:1 -x HCOLL_ENABLE_MCAST_ALL=0 -x HCOLL_MCAST_NP=5 -x HCOLL_CONTEXT_CACHE_ENABLE=0 -x UCX_SHM_DEVICES=all -x UCX_NET_DEVICES=mlx5_0:1 -x UCX_ACC_DEVICES=all -x HCOLL_ENABLE_SHARP=0 -x HCOLL_ENABLE_TOPOLOGY=0 -x HCOLL_BCOL_P2P_MCAST_ALLREDUCE_ALG=1 /hpc/scrap/users/mtt/scratch/hcol/20171016_215200_22084_17122_clx-orion-071/installs/gzmC/tests/mpi-test-suite/ompi-tests/mpi_test_suite/mpi_test_suite -x relaxed -t 'Alltoall' -d 'MPI_CONTIGUOUS_INT' -n 300`

Output:
```
Wed Oct 18 16:09:18 2017<stdout>:(Rank:0) tst_test_array[0]:Alltoall
Wed Oct 18 16:09:18 2017<stdout>:P2P tests Alltoall (23/1), comm MPI_COMM_WORLD (1/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:09:18 2017<stdout>:P2P tests Alltoall (23/1), comm Duplicated MPI_COMM_WORLD (4/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:09:18 2017<stdout>:[1508332158.841121] [clx-orion-046:7098 :0]    ib_mlx5_log.c:109  UCX  ERROR Error on QP 0x3b83f wqe[598]: Local length (synd 0x1 vend 0x68) opcode SEND
Wed Oct 18 16:09:18 2017<stdout>:[1508332158.841198] [clx-orion-046:7098 :0]     ucp_worker.c:380  UCX  ERROR Error Endpoint timeout was not handled for ep 0x189eae0 - dc_mlx5/mlx5_0:1
Wed Oct 18 16:09:18 2017<stdout>:[1508332158.875755] [clx-orion-046:7098 :0]        offload.c:476  UCX  ERROR Failed to cancel tag rndv op Endpoint timeout
Wed Oct 18 16:09:18 2017<stdout>:P2P tests Alltoall (23/1), comm Reversed MPI_COMM_WORLD (5/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:09:18 2017<stddiag>:[clx-orion-046:07098] pml_ucx.c:714 Error: ucx send failed: Endpoint timeout
[clx-orion-046:07098] *** An error occurred in MPI_Isend
[clx-orion-046:07098] *** reported by process [3194683393,140733193388366]
[clx-orion-046:07098] *** on communicator MPI COMMUNICATOR 4 CREATE FROM 0
[clx-orion-046:07098] *** MPI_ERR_OTHER: known error not in list
[clx-orion-046:07098] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[clx-orion-046:07098] ***    and potentially your MPI job)
```

MXM works:
```
$mpirun -np 1000 --bind-to core --map-by node -mca pml yalla -mca btl_openib_warn_default_gid_prefix 0 -mca btl_openib_if_include mlx5_0:1 --timestamp-output -x HCOLL_IB_IF_INCLUDE=mlx5_0:1 -x MXM_RDMA_PORTS=mlx5_0:1 -x HCOLL_ENABLE_MCAST_ALL=0 -x HCOLL_MCAST_NP=5 -x HCOLL_CONTEXT_CACHE_ENABLE=0 -x UCX_SHM_DEVICES=all -x UCX_NET_DEVICES=mlx5_0:1 -x UCX_ACC_DEVICES=all -x HCOLL_ENABLE_SHARP=0 -x HCOLL_ENABLE_TOPOLOGY=0 -x HCOLL_BCOL_P2P_MCAST_ALLREDUCE_ALG=1 /hpc/scrap/users/mtt/scratch/hcol/20171016_215200_22084_17122_clx-orion-071/installs/gzmC/tests/mpi-test-suite/ompi-tests/mpi_test_suite/mpi_test_suite -x relaxed -t 'Alltoall' -d 'MPI_CONTIGUOUS_INT' -n 300
Wed Oct 18 16:09:40 2017<stdout>:(Rank:0) tst_test_array[0]:Alltoall
Wed Oct 18 16:09:40 2017<stdout>:P2P tests Alltoall (23/1), comm MPI_COMM_WORLD (1/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:09:40 2017<stdout>:P2P tests Alltoall (23/1), comm Duplicated MPI_COMM_WORLD (4/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:09:41 2017<stdout>:P2P tests Alltoall (23/1), comm Reversed MPI_COMM_WORLD (5/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:09:41 2017<stdout>:P2P tests Alltoall (23/1), comm Halved MPI_COMM_WORLD (6/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:09:41 2017<stdout>:P2P tests Alltoall (23/1), comm Odd/Even split MPI_COMM_WORLD (7/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:09:41 2017<stdout>:P2P tests Alltoall (23/1), comm Zero-and-Rest Intercommunicator (8/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:09:41 2017<stdout>:P2P tests Alltoall (23/1), comm Halved Intercommunicator (12/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:09:41 2017<stdout>:P2P tests Alltoall (23/1), comm Intracomm merged of the Halved Intercomm (13/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:09:41 2017<stdout>:P2P tests Alltoall (23/1), comm MPI_COMM_TYPE_SHARED comm (14/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:09:41 2017<stdout>:Number of failed tests:0
```

Also reproduced without HCOLL:
```
$mpirun -np 1000 --bind-to core --map-by node -mca pml ucx -mca coll ^hcoll -mca btl_openib_warn_default_gid_prefix 0 -mca btl_openib_if_include mlx5_0:1 --timestamp-output -x HCOLL_IB_IF_INCLUDE=mlx5_0:1 -x MXM_RDMA_PORTS=mlx5_0:1 -x HCOLL_ENABLE_MCAST_ALL=0 -x HCOLL_MCAST_NP=5 -x HCOLL_CONTEXT_CACHE_ENABLE=0 -x UCX_SHM_DEVICES=all -x UCX_NET_DEVICES=mlx5_0:1 -x UCX_ACC_DEVICES=all -x HCOLL_ENABLE_SHARP=0 -x HCOLL_ENABLE_TOPOLOGY=0 -x HCOLL_BCOL_P2P_MCAST_ALLREDUCE_ALG=1 /hpc/scrap/users/mtt/scratch/hcol/20171016_215200_22084_17122_clx-orion-071/installs/gzmC/tests/mpi-test-suite/ompi-tests/mpi_test_suite/mpi_test_suite -x relaxed -t 'Alltoall' -d 'MPI_CONTIGUOUS_INT' -n 300
Wed Oct 18 16:10:04 2017<stdout>:(Rank:0) tst_test_array[0]:Alltoall
Wed Oct 18 16:10:04 2017<stdout>:P2P tests Alltoall (23/1), comm MPI_COMM_WORLD (1/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:10:04 2017<stdout>:P2P tests Alltoall (23/1), comm Duplicated MPI_COMM_WORLD (4/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:10:04 2017<stdout>:P2P tests Alltoall (23/1), comm Reversed MPI_COMM_WORLD (5/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:10:04 2017<stdout>:P2P tests Alltoall (23/1), comm Halved MPI_COMM_WORLD (6/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:10:04 2017<stdout>:P2P tests Alltoall (23/1), comm Odd/Even split MPI_COMM_WORLD (7/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:10:04 2017<stdout>:P2P tests Alltoall (23/1), comm Zero-and-Rest Intercommunicator (8/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:10:04 2017<stdout>:P2P tests Alltoall (23/1), comm Halved Intercommunicator (12/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:10:04 2017<stdout>:P2P tests Alltoall (23/1), comm Intracomm merged of the Halved Intercomm (13/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:10:04 2017<stdout>:[1508332204.790096] [clx-orion-082:5228 :0]    ib_mlx5_log.c:109  UCX  ERROR Error on QP 0xfbe4 wqe[1516]: Local QP operation (synd 0x2 vend 0x68) opcode SEND
Wed Oct 18 16:10:04 2017<stdout>:[1508332204.790188] [clx-orion-082:5228 :0]     ucp_worker.c:380  UCX  ERROR Error Endpoint timeout was not handled for ep 0x141d510 - dc_mlx5/mlx5_0:1
Wed Oct 18 16:10:04 2017<stdout>:[1508332204.800257] [clx-orion-082:5228 :0]        offload.c:476  UCX  ERROR Failed to cancel tag rndv op Endpoint timeout
Wed Oct 18 16:10:04 2017<stdout>:P2P tests Alltoall (23/1), comm MPI_COMM_TYPE_SHARED comm (14/14), type MPI_CONTIGUOUS_INT (21/1)
Wed Oct 18 16:10:04 2017<stdout>:Number of failed tests:0
```

### PR DESCRIPTION
**Pull request by yosefe**
Fixes #1926

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6336/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6339/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6340/ for details (Mellanox internal link).

File: src/uct/ib/rc/base/rc_iface.c [hunk 1]
```diff
@@ -919,7 +920,7 @@ ucs_status_t uct_rc_iface_qp_create(uct_rc_iface_t *iface, int qp_type,
 
 #  if HAVE_DECL_IBV_EXP_ATOMIC_HCA_REPLY_BE
     if (dev->dev_attr.exp_atomic_cap == IBV_EXP_ATOMIC_HCA_REPLY_BE) {
-        qp_init_attr.comp_mask       |= IBV_EXP_QP_INIT_ATTR_CREATE_FLAGS;
+        qp_init_attr.comp_mask      |= IBV_EXP_QP_INIT_ATTR_CREATE_FLAGS;
```
Comment (brminich) [inline src/uct/ib/rc/base/rc_iface.c]: previous alignment was correct

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 2]
```diff
@@ -236,7 +236,8 @@ static UCS_CLASS_INIT_FUNC(uct_rc_mlx5_iface_t, uct_md_h md, uct_worker_h worker
 
     UCS_CLASS_CALL_SUPER_INIT(uct_rc_iface_t, &uct_rc_mlx5_iface_ops, md, worker,
                               params, &config->super, 0,
-                              sizeof(uct_rc_fc_request_t), IBV_EXP_TM_CAP_RC);
+                              sizeof(uct_rc_fc_request_t), IBV_EXP_TM_CAP_RC,
+                              UCT_IB_MLX5_RES_DOMAIN_KEY);
```
Comment (brminich) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c]: Is it ok that all accelerated transports use the same domain key?

**Comment by shamisp** [comment]
ouch... this is critical one.

BTW, Can we add some check (ucx info ?) validating that majority of BF doorbells are successful ?

**Comment by yosefe** [comment]
reading the counters is not possible without root privileges..

**Comment by shamisp** [comment]
yes...

File: src/uct/ib/base/ib_iface.c [hunk 3]
```diff
@@ -534,6 +534,60 @@ static ucs_status_t uct_ib_iface_set_moderation(struct ibv_cq *cq,
...
 }
 
+static int uct_ib_iface_res_domain_cmp(uct_ib_iface_res_domain_t *res_domain,
+                                       uct_ib_iface_t *iface)
+{
+    uct_ib_device_t *dev = uct_ib_iface_device(iface);
+
+    return res_domain->ibv_domain->context == dev->ibv_context;
+}
+
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/base/ib_iface.c]: fixed

File: src/uct/ib/rc/base/rc_iface.c (hunk 1 above)
Comment (yosefe) [reply src/uct/ib/rc/base/rc_iface.c]: fixed

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 2 above)
Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c]: yes because they share the BF register structure as well (with the same key)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6491/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6499/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6600/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6610/ for details (Mellanox internal link).

File: config/m4/ib.m4 [hunk 4]
```diff
@@ -164,6 +164,16 @@ AS_IF([test "x$with_ib" == xyes],
                                [[#include <infiniband/mlx5_hw.h>]])],
              [])
 
+       AC_CHECK_DECLS([IBV_EXP_QP_INIT_ATTR_RES_DOMAIN,
+                       IBV_EXP_RES_DOMAIN_THREAD_MODEL,
+                       ibv_exp_create_res_domain,
+                       ibv_exp_destroy_res_domain],
+                      [AC_DEFINE([HAVE_IBV_EXP_RES_DOMAIN], 1, [IB resource domain])],
+                      [AC_MSG_WARN([Cannot use mlx5 accel because resource domains are not supported])
+                       AC_MSG_WARN([Please upgrade MellanoxOFED to 3.1 or above])
```
Comment (shamisp) [inline config/m4/ib.m4]: add URL

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6612/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:bgate:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6629/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6631/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@brminich pls take a look

**Pull request merged**